File: test-suite-java/src/test/java/io/micronaut/sourcegen/example/MyBean1Test.java
/*
 * Copyright 2003-2021 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.micronaut.sourcegen.example;

import org.junit.jupiter.api.Test;

import java.lang.reflect.Modifier;
import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertInstanceOf;
import static org.junit.jupiter.api.Assertions.assertTrue;

class MyBean1Test {
    @Test
    public void test() throws Exception {
        MyBean1 bean = new MyBean1();
        bean.setId(123);
        bean.setName("TheName");
        bean.setAge(55);
        bean.setAddresses(List.of("Address 1"));
        bean.setTags(List.of("X", "Y"));

        assertEquals("TheName", bean.getName());
        assertEquals(123, bean.getId());
        assertEquals(55, bean.getAge());
        assertEquals(List.of("Address 1"), bean.getAddresses());
        assertEquals(List.of("X", "Y"), bean.getTags());

        bean.setId(987);
        bean.setName("Xyz");
        bean.setAge(123);
        bean.setAddresses(List.of("Address X", "Address Y"));
        bean.setTags(List.of("A", "B", "C"));

        assertEquals("Xyz", bean.getName());
        assertEquals(987, bean.getId());
        assertEquals(123, bean.getAge());
        assertEquals(List.of("Address X", "Address Y"), bean.getAddresses());
        assertEquals(List.of("A", "B", "C"), bean.getTags());

        assertTrue(Modifier.isPrivate(
            bean.getClass().getDeclaredField("id").getModifiers()
        ));
        assertInstanceOf(Deprecated.class, bean.getClass().getDeclaredField("id").getDeclaredAnnotations()[0]);
        assertTrue(Modifier.isPublic(
            bean.getClass().getDeclaredMethod("getId").getModifiers()
        ));
        Deprecated deprecated = (Deprecated) bean.getClass().getDeclaredField("age").getDeclaredAnnotations()[0];
        assertEquals(deprecated.since(), "xyz");
        assertTrue(deprecated.forRemoval());
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/sourcegen/example/MyBean1Test.kt
package io.micronaut.sourcegen.example

import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test
import java.lang.Deprecated
import java.lang.reflect.Modifier

class MyBean1Test {

    @Test
    @Throws(Exception::class)
    fun test() {
        val bean = MyBean1(123, "TheName", 55, listOf(), listOf())
        assertEquals("TheName", bean.name)
        assertEquals(123, bean.id)
        assertEquals(55, bean.age)
        bean.name = "Xyz"
        bean.id = 987
        bean.age = 123
        bean.addresses = listOf("Address 1", "Address 2")
        bean.tags = listOf("Tag 1", "Tag 2")
        assertEquals("Xyz", bean.name)
        assertEquals(987, bean.id)
        assertEquals(123, bean.age)
        assertEquals(listOf("Address 1", "Address 2"), bean.addresses)
        assertEquals(listOf("Tag 1", "Tag 2"), bean.tags)

        Assertions.assertTrue(
            Modifier.isPrivate(
                bean.javaClass.getDeclaredField("id").modifiers
            )
        )
        Assertions.assertFalse(
            Modifier.isFinal(
                bean.javaClass.getDeclaredField("id").modifiers
            )
        )
        // For primary constructor bean the annotation are set on the constructor parameters
        Assertions.assertTrue(
            bean.javaClass.constructors[0].parameters[0].declaredAnnotations[0] is Deprecated
        )
        Assertions.assertTrue(
            Modifier.isPublic(
                bean.javaClass.getDeclaredMethod("getId").modifiers
            )
        )
        Assertions.assertNotNull(
                bean.javaClass.declaredMethods.find { it.name == "setId" }
        )
        val deprecated = bean.javaClass.constructors[0].parameters[2].declaredAnnotations[0] as Deprecated
        assertEquals(deprecated.since, "xyz")
        Assertions.assertTrue(deprecated.forRemoval)
    }
}

